### PR TITLE
Fix treatment of Vec of Analog and Vec of Bundle of Analog

### LINKED ISSUE #1083: Vec of Analog is broken
```scala
import chisel3._
import chisel3.experimental.{RawModule, Analog}                                                                                                                                   

class MyModule extends RawModule {
  val bus = IO(Vec(4, Analog(1.W)))
}

object TestMain extends App {
  println(chisel3.Driver.emitVerilog(new MyModule))
}
```

Chisel errors:
```
[info] [info] [0.008] Elaborating design...
[error] Exception in thread "main" scala.MatchError: SampleElementBinding(Analog<1>[4](IO in unelaborated MyModule)) (of class chisel3.core.SampleElementBinding)
[error]         at chisel3.core.Analog.bind(Bits.scala:1776)
[error]         at chisel3.core.Vec.bind(Aggregate.scala:143)
[error]         at chisel3.core.BaseModule._bindIoInPlace(Module.scala:315)
[error]         at chisel3.core.BaseModule.bindIoInPlace(Module.scala:319)
[error]         at chisel3.core.IO$.apply(Module.scala:114)
[error]         at chisel3.core.BaseModule.IO(Module.scala:337)
[error]         at MyModule.<init>(Test.scala:6)
```

I bisected and found that this is my fault 😞 #946 

**Type of issue**: bug report

**Impact**: no functional change

**Development Phase**: request

**Other information**
 

**If the current behavior is a bug, please provide the steps to reproduce the problem:**

**What is the current behavior?**

**What is the expected behavior?**

**Please tell us about your environment:**

**What is the use case for changing the behavior?**

### PR DESCRIPTION
**Pull request by jackkoenig**
Refactoring of https://github.com/freechipsproject/chisel3/pull/1071, thank you @ckdur

I simplified the test a little but, and fixed the addition case of `IO(Vec(n, new Bundle { val x = Analog(8.W) }))`. I added `Data.isSynthesizable` to distinguish the behavior of SampleElementBinding in that vs. `Data.topBindingOpt`. This helps clarify the notion of being bound but not hardware.
Data.topBindingOpt is now used to get the *actual* top binding,
including across SampleElements (eg. in Analog checking that the top is
bound to a Port or a Wire).

I then did an audit of all uses of `topBindingOpt` to see which were actually checking if the thing is hardware and replaced them with that. The only other modification I had to make was the pretty-printing of Vec since it uses `topBindingOpt` and was getting the `(Wire in ...)` part twice.

Please take a look, we need this in rocket-chip asap because #1083 is causing pain in projects built on top of it.

Fixes #1083 

**Related issue**: #946 

**Type of change**: bug fix

**Impact**: no functional change

**Development Phase**: implementation

**Release Notes**

### CODE REVIEW

**Comment by colinschmidt** [comment]
New tests look good to me.

File: chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala [hunk 1]
```diff
@@ -136,7 +136,8 @@ object Vec extends VecFactory
 sealed class Vec[T <: Data] private[core] (gen: => T, val length: Int)
     extends Aggregate with VecLike[T] {
   override def toString: String = {
-    s"$sample_element[$length]$bindingToString"
+    val elementType = sample_element.cloneType
```
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala]: Wait why is this necessary? This smells odd...

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala [hunk 2]
```diff
@@ -314,14 +314,18 @@ abstract class Data extends HasId with NamedComponent with SourceInfoDoc { // sc
     _binding = Some(target)
   }
 
+  // Similar to topBindingOpt except it explicitly excludes SampleElements which are bound but not
+  // hardware
+  private[core] final def isSynthesizable: Boolean = _binding.map {
```
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Currently we define properties of the different bindings in the Binding type hierarchy, can we introduce a SynthesizableBinding trait and check for that, instead of scattering binding definition logic throughout?

File: chiselFrontend/src/main/scala/chisel3/core/RawModule.scala [hunk 3]
```diff
@@ -75,12 +75,14 @@ abstract class RawModule(implicit moduleCompileOptions: CompileOptions)
...
-        case id: Data if id.topBindingOpt.isDefined => id.topBinding match {
-          case OpBinding(_) | MemoryPortBinding(_) | PortBinding(_) | RegBinding(_) | WireBinding(_) =>
-            id.forceName(default="_T", _namespace)
-          case _ =>  // don't name literals
-        }
-        case id: Data if id.topBindingOpt.isEmpty =>  // don't name unbound types
+        case id: Data  =>
+          if (id.isSynthesizable) {
+            id.topBinding match {
+              case OpBinding(_) | MemoryPortBinding(_) | PortBinding(_) | RegBinding(_) | WireBinding(_) =>
```
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/RawModule.scala]: I know this code as previously structured like this but is this something that can also be factored into binding traits?

Comment (jackkoenig) [reply chiselFrontend/src/main/scala/chisel3/core/RawModule.scala]: I think such refactoring makes sense, but the main purpose of this PR is to fix a bug

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala (hunk 2 above)
Comment (jackkoenig) [reply chiselFrontend/src/main/scala/chisel3/core/Data.scala]: I thought about such a refactoring but it seems like a large task for a bug fix, we need this bug fixed to bump rocket-chip, and the code is strictly more explicit than it was before

File: chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala (hunk 1 above)
Comment (jackkoenig) [reply chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala]: Because [`bindingToString`](https://github.com/freechipsproject/chisel3/blob/c1ab9e7afd5072c11d879db913e1b553c7fe0dbe/chiselFrontend/src/main/scala/chisel3/core/Data.scala#L354) uses `topBindingOpt`. I changed `topBindingOpt` to return the actual top binding (which you want sometimes from a `sample_element`). So in order to keep the pretty printing sane: `UInt<8>[8](Wire in ...)` rather than `UInt<8>(Wire in ...)[8](Wire in ...)`, I thought it prudent to clone it to clear the binding.

Comment (ducky64) [reply chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala]: An alternative would be to make SampleElementBinding a TopBinding (it kind of is, at least in that it's not a hardware child of the container Vec), and have bindingToString for SampleElementBinding return something sane?

Comment (jackkoenig) [reply chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala]: I was thinking about that and I think it's a reasonable approach but there are a couple of issues that seem problematic (or at least weird) to me:

1. `SampleElementBinding` isn't truly a `TopBinding`, because it still needs to point to the `Vec` of which it's the sample element. One reason why this is true is discussed below,

2. We need to check that Analog is only bound to `Port`s and `Wire`s, not `Reg`s or `Mem`s. From the code:

```scala
    val targetTopBinding = target match {
      case target: TopBinding => target
      case ChildBinding(parent) => parent.topBinding
      // See https://github.com/freechipsproject/chisel3/pull/946
      case SampleElementBinding(parent) => parent.topBinding
    }

    // Analog counts as different directions based on binding context
    targetTopBinding match {
      case WireBinding(_) => direction = ActualDirection.Unspecified  // internal wire
      case PortBinding(_) => direction = ActualDirection.Bidirectional(ActualDirection.Default)
      case x => throwException(s"Analog can only be Ports and Wires, not '$x'")
    }
```
If we make `SampleElementBinding` a pseudo-`TopBinding`, we could rewrite it as something like:
```scala
    val targetTopBinding = target match {
      case SampleElementBinding(parent) => parent.topBinding // <-- Note this is also a TopBinding so has to be first
      case target: TopBinding => target
      case ChildBinding(parent) => parent.topBinding  
    }

    // Analog counts as different directions based on binding context
    targetTopBinding match {
      case WireBinding(_) => direction = ActualDirection.Unspecified  // internal wire
      case PortBinding(_) => direction = ActualDirection.Bidirectional(ActualDirection.Default)
      case x => throwException(s"Analog can only be Ports and Wires, not '$x'")
    }
```
We can't rely upon the non-sample elements for this check, because then the following (which should error) would not error when n == 0:
```scala
  val reg = Reg(Vec(n, Analog(1.W))
```

Comment (jackkoenig) [reply chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala]: I agree that the `cloneType` smells weird, but it smells less weird to me than the existence of `SampleElement` as only a pseudo-`TopBinding`

Comment (jackkoenig) [reply chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala]: Actually I think worse than what I said above because you could have nested Vecs as your type and thus you hit multiple `SampleElement` bindings on your way, so you'd have to do something like:

```scala
    val targetTopBinding = {
      def rec(binding: Binding): TopBinding = binding match {
        case SampleElementBinding(parent) => rec(parent.topBinding)
        case target: TopBinding => target
        case ChildBinding(parent) => rec(parent.topBinding)
      }
      rec(target)
    }
```
This makes it seem to me that `SampleElementBinding` doesn't make sense as a `TopBinding`.

Comment (ducky64) [reply chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala]: Yeah, this seems like a thorny question. I'm fine merging this hotfix as-is, and perhaps we should have a separate discussion on SampleElementBinding in its own issue.

**Comment by aswaterman** [comment]
OK, I'm going to push this through, and we can deal with the cleanups in separate PRs.
